exporters/prometheus/exporter.go

```diff
@@ -0,0 +1,185 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package prometheus // import "github.com/open-telemetry/opentelemetry-go/exporters/prometheus"
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+// Exporter is a Prometheus exporter.
+type Exporter struct {
+	reader   metric.Reader
+	registry *prometheus.Registry
+}
+
+// NewPrometheusExporter returns a Prometheus exporter.
+func NewPrometheusExporter(reader metric.Reader) (*Exporter, error) {
+	e := &Exporter{
+		reader:   reader,
+		registry: prometheus.NewRegistry(),
+	}
+
+	if err := e.registry.Register(e); err != nil {
+		return nil, err
+	}
+
+	return e, nil
+}
+
+// Describe implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
```

Comment:
Here is a suggestion (non-blocking) for how to simplify this:

* Describe() calls Collect(), iterates over the Metrics in the collect channel, and calls [Desc()](https://github.com/prometheus/client_golang/blob/3faf3bae7076031e56adc1f76d0c2d817410597e/prometheus/metric.go#L41) on each metric, and puts it in the channel.
* Now, you just need to implement Collect (not describe), so you can get rid of metricData, and just pass the `ch` channel from collect into getMetricData.

Reply:
could you show what you mean by non-blocking? Using a switch won't iterate over all the values but ranging over the collect channel in Describe() means I have to close that channel at the end of Collect(). Having trouble with this because the client_golang registry code [also closes those channels](https://github.com/prometheus/client_golang/blob/v1.13.0/prometheus/registry.go#L471-L472), so I end up with a panic:
```
panic: close of closed channel

goroutine 24 [running]:
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func2()
	/usr/local/google/home/mikedame/go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/registry.go:472 +0x45
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
	/usr/local/google/home/mikedame/go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/registry.go:469 +0x5db
FAIL	go.opentelemetry.io/otel/exporters/prometheus	0.041s
FAIL
```

Reply:
> could you show what you mean by non-blocking? 

I just mean that you can do it as a follow-up (the comment doesn't block merging).

Reply:
Ah thought you meant literally a non-blocking channel call, hah

Reply:
You can use the DescribeByCollect helper to do this for you: https://github.com/prometheus/client_golang/blob/5b7e8b2e6716df0ceda9df81feb15910c7efa150/prometheus/collector.go#L87